lib/proto/proto.go

```diff
@@ -969,6 +1040,145 @@ func (it *repeatedFieldIterator) Done() {
 	}
 }
 
+type MapField struct {
+	typ protoreflect.FieldDescriptor
+
+	mp        protoreflect.Map
+	frozen    *bool
+	itercount int
+}
+
+var _ starlark.HasSetKey = (*MapField)(nil)
+var _ starlark.IterableMapping = (*MapField)(nil)
+
+func (mf *MapField) Type() string {
+	return fmt.Sprintf("proto.map<%s, %s>", typeString(mf.typ.MapKey()), typeString(mf.typ.MapValue()))
+}
+
+func (mf *MapField) SetKey(k, v starlark.Value) error {
+	if err := mf.checkMutable("set key in"); err != nil {
+		return err
+	}
+
+	kx, err := toProto(mf.typ.MapKey(), k)
+	if err != nil {
+		return fmt.Errorf("setting key of map field: %v", err)
+	}
+	vx, err := toProto(mf.typ.MapValue(), v)
+	if err != nil {
+		return fmt.Errorf("setting value of map field: %v", err)
+	}
+
+	mf.mp.Set(kx.MapKey(), vx)
+	return nil
+}
+
+// checkMutable reports an error if the map field should not be mutated.
+// verb+" map field" should describe the operation.
+func (mf *MapField) checkMutable(verb string) error {
+	if *mf.frozen {
+		return fmt.Errorf("cannot %s frozen map field", verb)
+	}
+	if mf.itercount > 0 {
+		return fmt.Errorf("cannot %s map field during iteration", verb)
+	}
+	return nil
+}
+
+func (mf *MapField) Get(k starlark.Value) (starlark.Value, bool, error) {
+	pk, err := toProto(mf.typ.MapKey(), k)
+	if err != nil {
+		return nil, false, fmt.Errorf("getting value of map field: %v", err)
+	}
+
+	v := mf.mp.Get(pk.MapKey())
+	if !v.IsValid() {
+		return nil, false, nil
+	}
+
+	return toStarlark1(mf.typ.MapValue(), v, mf.frozen), true, nil
+}
+
+func (mf *MapField) Freeze()               { *mf.frozen = true }
+func (mf *MapField) Hash() (uint32, error) { return 0, fmt.Errorf("unhashable: %s", mf.Type()) }
+
+func (mf *MapField) Iterate() starlark.Iterator {
+	if !*mf.frozen {
+		mf.itercount++
+	}
+
+	// TODO(negz): Should we store only keys in the iterator? It doesn't
+	// consume the values.
+	return &mapFieldIterator{mf, mf.Items(), 0}
+}
+
+func (mf *MapField) Items() []starlark.Tuple {
+	out := make([]starlark.Tuple, 0, mf.mp.Len())
+
+	mf.mp.Range(func(mk protoreflect.MapKey, v protoreflect.Value) bool {
+		out = append(out, starlark.Tuple{
+			toStarlark1(mf.typ.MapKey(), mk.Value(), mf.frozen),
+			toStarlark1(mf.typ.MapValue(), v, mf.frozen),
+		})
+		return true // Keep iterating.
+	})
+
+	// A map key can be any scalar protobuf type except floats and bytes.
+	// In practice in starlark they should be Int, String, or Bool and thus
+	// either TotallyOrdered or Comparable. None of these return errors when
+	// compared to the same type.
+	sort.Slice(out, func(i, j int) bool {
+		less, _ := starlark.Compare(syntax.LT, out[i][0], out[j][0])
+		return less
+	})
+
+	return out
+}
+
+func (mf *MapField) Len() int { return mf.mp.Len() }
+
+func (mf *MapField) String() string {
+	// We want to use {k1: v1, k2: v2} notation, like a dict.
+	buf := new(bytes.Buffer)
```

Comment:
Use strings.Builder to save an allocation, since the result is desired as a string.